### PR TITLE
chore: fix "Invalid example" warnings from shred map builder

### PR DESCRIPTION
**Pull request by chalin**
The shred map (xref) builder was issuing warnings. This fix includes
- Adjustments to the shredder map builder itself so that it
understands, e.g., app-project relative example paths.
- `**/guide/glossary.jade` now (Jade) `includes` the shared parent
`glossary.jade` rather than (Harp) importing (via `partial`). This
fixes `makeExample` path issues in the glossary.
- Adjusted some `makeExample` paths that were ok for site build, but
confused the xref tool.

### CODE REVIEW

**Comment by chalin** [comment]
@Foxandxss @wardbell @kwalrath : ready for review.

Here are the ngio docs related warnings:
```
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/dependency-injection/ts/app/app.config-config-alt.ts.md") - doc - from file "dart/latest/guide/dependency-injection.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/toh-5/dart/web/index-router.html.md") - doc - from file "dart/latest/tutorial/toh-pt5.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/app/in-memory-data.service-init.ts.md") - doc - from file "ts/latest/tutorial/toh-pt6.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/app/heroes.component-delete.html.md") - doc - from file "ts/latest/tutorial/toh-pt6.jade"
```

And here are the ng API doc example warnings (that were repeated for TS):
```
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/platform/dom/debug/ts/by/by-by_all.ts.md") - doc - from file "js/latest/api/platform-browser/index/By-class.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/platform/dom/debug/ts/by/by-by_css.ts.md") - doc - from file "js/latest/api/platform-browser/index/By-class.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/platform/dom/debug/ts/by/by-by_directive.ts.md") - doc - from file "js/latest/api/platform-browser/index/By-class.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/can_activate/can_activate_example-canActivate.ts.md") - doc - from file "js/latest/api/router-deprecated/index/CanActivate-decorator.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/can_deactivate/can_deactivate_example-routerCanDeactivate.ts.md") - doc - from file "js/latest/api/router-deprecated/index/CanDeactivate-interface.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/reuse/reuse_example-reuseCmp.ts.md") - doc - from file "js/latest/api/router-deprecated/index/CanReuse-interface.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/on_activate/on_activate_example-routerOnActivate.ts.md") - doc - from file "js/latest/api/router-deprecated/index/OnActivate-interface.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/on_deactivate/on_deactivate_example-routerOnDeactivate.ts.md") - doc - from file "js/latest/api/router-deprecated/index/OnDeactivate-interface.jade"
warn:    Invalid example (unable to locate fragment file: "public/docs/_fragments/_api/router/ts/reuse/reuse_example-reuseCmp.ts.md") - doc - from file "js/latest/api/router-deprecated/index/OnReuse-interface.jade"
```

The latter have been fixed via https://github.com/angular/angular/pull/9828.

**Comment by Foxandxss** [comment]
So this fix the two ToH issues with fragments? Those are fine on the .jade.

**Comment by chalin** [comment]
Yes it fixes the toh fragment issues. And you are correct: those were fine in the rendered site HTML. The problem arose in the fragment cross referencer which Ward said we will likely be using once again in the future (cf. [this](https://angular-team.slack.com/archives/docs-authoring/p1467747319001559) comment).

File: tools/doc-shredder/processors/shredMapProcessor.js [hunk 1]
```diff
@@ -18,25 +18,37 @@ module.exports = function shredMapProcessor(log, createDocMessage) {
...
           }
-          var region = fragItem.region ? "-" + fragItem.region : '';
-          var extn = path.extname(mixinPath);
-          var basename = path.basename(mixinPath, extn);
-          var fragDir = path.dirname(mixinPath);
-          var fragPath = path.join(fragDir, basename + region + extn) + '.md';
-          var fullFragPath =  path.join(options.fragmentsDir, fragPath);
-
-          var fragInfo = { fragPath: fullFragPath, examplePath: fullExamplePath, exists: fs.existsSync(fullFragPath) };
+          var fragInfo = makeFragInfo(options.fragmentsDir, fullExamplePath, fragItem, mixinPath);
```
Comment (wardbell) [inline tools/doc-shredder/processors/shredMapProcessor.js]: Can you explain what the consequences of this (and similar changes) are for the way we write examples in Jade?

File: public/docs/ts/latest/glossary.jade [hunk 2]
```diff
@@ -54,7 +54,7 @@ include _util-fns
     The barrel itself is a module file that re-exports *selected* exports of other modules.
     
     Imagine three modules in a `heroes` folder:
-  code-example(format='').
+  code-example.
```
Comment (wardbell) [inline public/docs/ts/latest/glossary.jade]: Why removing `(format='')` here and elsewhere in this file?
Not against it. Just wondering why the change.

File: tools/doc-shredder/processors/shredMapProcessor.js [hunk 3]
```diff
@@ -82,13 +94,24 @@ module.exports = function shredMapProcessor(log, createDocMessage) {
...
-  var rx = /(.*)\-(.*)\.(.s)/;
-  var r = rx.exec(fragPath);
-  if (r) {
-    return r[1] + '.' + r[3];
-  } else {
-    return fragPath;
-  }
+// TODO: use the functionality in public/resources/js/util.js once it lands.
+function jadeBaseFileNameToExampleName(name) {
+  // Adjust for known cases where chapter name is not the example name.
```
Comment (wardbell) [inline tools/doc-shredder/processors/shredMapProcessor.js]: Kind of a PITA I know.  I wanted to fix the folders so we didn't have the discrepancy that requires this hack. That would mean adding firebase rewrite rules to handle external links. Sigh. I guess this gets it done. But it adds technical debt.

File: public/docs/ts/latest/glossary.jade (hunk 2 above)
Comment (chalin) [inline public/docs/ts/latest/glossary.jade]: Short answer: because it has no effect `code-example` (see the [source](https://github.com/angular/angular.io/blob/master/public/resources/js/directives/code-example.js)).

Longer answer: this idiom comes from the `makeExample` mixin (where it prevents the default behavior of turning on line numbers if there is more than one line). But is has no utility when applied directly to `code-example`.

File: tools/doc-shredder/processors/shredMapProcessor.js (hunk 1 above)
Comment (chalin) [inline tools/doc-shredder/processors/shredMapProcessor.js]: It has no impact on how we write the Jade, but it allows the shred xref mapper to make sense of the abbreviated `makeExample` source paths like:
```jade
+makeExample('app/hero.service.ts')
```
vs the original long form:
```jade
+makeExample('router/ts/app/main.1.ts','all', 'main.ts')(format=".")
```

Does that answer your question?

File: tools/doc-shredder/processors/shredMapProcessor.js (hunk 3 above)
Comment (chalin) [inline tools/doc-shredder/processors/shredMapProcessor.js]: I agree.

File: public/docs/ts/latest/glossary.jade (hunk 2 above)
Comment (wardbell) [inline public/docs/ts/latest/glossary.jade]: Actually, something has changed. `code-example` ***used to*** display line numbers by default. 

Now the default is *no line numbers*. If you want them you have to add them explicitly as in:
```
code-example(format="linenums").
  // Code with line numbers
  one
  3
  4
```
I wonder when this change took place? Not that I'm objecting.

**Comment by wardbell** [comment]
Merged
